components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java

```diff
@@ -0,0 +1,149 @@
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.Response;
+
+import static org.wso2.carbon.identity.oauth.endpoint.util.EndpointUtil.getOAuth2Service;
+
+/**
+ * REST implementation for OAuth2 PAR endpoint.
+ * The endpoint accepts POST request with the authorization parameters.
+ * Returns a request_uri as a reference for the submitted parameters and the expiry time.
+ */
+@Path("/par")
+@InInterceptors(classes = OAuthClientAuthenticatorProxy.class)
+public class OAuth2ParEndpoint {
+
+    @POST
+    @Path("/")
+    @Consumes("application/x-www-form-urlencoded")
+    @Produces("application/json")
+    public Response par(@Context HttpServletRequest request, @Context HttpServletResponse response,
+                        MultivaluedMap<String, String> params) {
+
+        try {
+            handleValidation(request, params);
+            Map<String, String> parameters = transformParams(params);
+            ParAuthResponseData parAuthResponseData =
+                    EndpointUtil.getParAuthService().generateParAuthResponse(parameters);
+            return createAuthResponse(response, parAuthResponseData);
+        } catch (ParClientException e) {
+            return handleParClientException(e);
+        } catch (ParCoreException e) {
+            return handleParCoreException(e);
+        }
+    }
+
+    private Map<String, String> transformParams(MultivaluedMap<String, String> params) {
+
+        Map<String, String> parameters = new HashMap<>();
+        for (Map.Entry<String, List<String>> entry : params.entrySet()) {
+            String key = entry.getKey();
+            List<String> values = entry.getValue();
+            if (!values.isEmpty()) {
+                String value = values.get(0);
+                parameters.put(key, value);
+            }
+        }
+
+        return parameters;
+    }
+
+    private Response createAuthResponse(HttpServletResponse response, ParAuthResponseData parAuthResponseData) {
+
+        response.setContentType(MediaType.APPLICATION_JSON);
+        net.minidev.json.JSONObject parAuthResponse = new net.minidev.json.JSONObject();
+        parAuthResponse.put(OAuthConstants.OAuth20Params.REQUEST_URI,
+                ParConstants.REQUEST_URI_HEAD + parAuthResponseData.getReqUriUUID());
+        parAuthResponse.put(ParConstants.EXPIRES_IN, parAuthResponseData.getExpiryTime());
+        Response.ResponseBuilder responseBuilder = Response.status(HttpServletResponse.SC_CREATED);
+        return responseBuilder.entity(parAuthResponse.toString()).build();
+    }
+
+    private Response handleParClientException(ParClientException exception) {
+
```

Comment:
Since this is the last place of handling the client error lets log this as a debug log